### PR TITLE
Update CDC500_states query

### PR DESCRIPTION
**Pull request by shvngisingh**
### Overview
Resolves issue b/530863116. Migrates `scripts/us_cdc/cdc500_state/process.py` from querying the unpartitioned `datcom-store.dc_kg_latest.StatVarObservation` table (which required a 3-way Cartesian self-join taking 2 hours 27 minutes / 8,823s and repeatedly timed out on Cloud Batch) to Spanner-backed BigQuery export tables (`TimeSeries` and `Observation` in `datcom-store.spanner_dc_graph_prod_DEFAULT`). This achieves a ~446x speedup (19.8s execution time) and eliminates historical multi-hierarchy double-counting bugs.

### Key Changes
1. **BigQuery Query Optimization (`process.py`)**:
   - Replaces unindexed full-table scans with indexed joins on `Observation` and `TimeSeries` pre-filtered by provenance (`dc/base/CDC500` and `dc/base/CensusACS5YearSurvey`).
   - Eliminates county + city + tract double-counting by restricting aggregation to 13-character city/CDP DCIDs (`LENGTH(O.entity1) = 13`), with a surgical exception for 11-character Honolulu County (`geoId/15003`) in `<= 2016` and `2017` non-BP/cholesterol indicators.
   - Maps percentage StatVars to exact Census ACS 5-Year demographic cohort denominators.
   - Excludes 4 non-standard multi-year cancer screening indicators (`Mammography`, `CervicalCancerScreening`, `PapSmearTest`, `ColorectalCancerScreening`) that lack composite Census ACS population denominators.
   - Deduplicates observations using `QUALIFY ROW_NUMBER() OVER (... ORDER BY O.last_update_timestamp DESC, O.facet_id DESC) = 1`.
   - Adds atomic write verification (`.tmp` write + `os.path.getsize > 0` check + `os.replace()`), project logging before dispatch, and `--timeout`, `--project`, and `--output_dir` flags.

2. **Unit Tests & Quality Gates (`process_test.py`)**:
   - Adds 11 hermetic unit tests executing `process.QUERY` in DuckDB across cohort mappings, Honolulu County 2016/2017/2018 edge cases, deduplication tie-breakers, atomic writes, and error handling (10.00/10 pylint score).

3. **Validation & Manifest (`validation_config.json`, `manifest.json`, `cdc500_state.tmcf`)**:
   - Implements scoped `SQL_VALIDATOR` (`check_statvar_max_dates`) enforcing cohort-specific vintage thresholds (`2016` for `WithAllTeethLoss`, `2020` for `CorePreventiveServices`, `2021` for `HighBloodPressure|Cholesterol|ChronicKidneyDisease`, `2022` for all others).
   - Enforces percentage bounds `[0, 100]` and place counts `[50, 52]`.
   - Right-sizes manifest resources to `cpu: 2, memory: 8, disk: 20` and registers `validation_config.json` in `source_files`.

### Test Environment Validation (`datcom-import-test`)
- **Cloud Batch Job ID**: `[link](https://pantheon.corp.google.com/batch/jobsDetail/regions/us-east4/jobs/cdc500-states-shvngisingh-20260918-112222/logs?project=datcom-infosys-dev)
- **Output GCS URI**: [link](https://pantheon.corp.google.com/storage/browser/datcom-import-test/scripts/us_cdc/cdc500_state/CDC500_States/2026_09_18T04_25_54_981182_07_00?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&forceOnBucketsSortingFiltering=true&project=datcom-infosys-dev)
- **Validation Results (`validation_output.csv`)**: [link](https://pantheon.corp.google.com/storage/browser/_details/datcom-import-test/scripts/us_cdc/cdc500_state/CDC500_States/2026_09_18T04_25_54_981182_07_00/input0/validation/validation_output.csv;tab=live_object?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&forceOnBucketsSortingFiltering=true&project=datcom-infosys-dev)

### CODE REVIEW

File: scripts/us_cdc/cdc500_state/process.py [hunk 1]
```diff
@@ -22,49 +22,70 @@
...
+    O.date AS observation_date,
+    O.value AS population
+  FROM `datcom-store.spanner_dc_graph_prod_DEFAULT.Observation` AS O
+  JOIN `datcom-store.spanner_dc_graph_prod_DEFAULT.TimeSeries` AS T
+    ON O.variable_measured = T.variable_measured
+    AND O.entity1 = T.entity1
+    AND O.facet_id = T.facet_id
+  WHERE O.entity1 LIKE 'geoId/%'
+    AND T.variable_measured IN (SELECT DISTINCT pop_statvar FROM cdc_sv)
+)
```
[the remark this answers is not in the document]
Comment (shvngisingh) [reply scripts/us_cdc/cdc500_state/process.py]: added change

File: scripts/us_cdc/cdc500_state/process.py [hunk 2]
```diff
@@ -22,49 +22,70 @@
...
+SELECT DISTINCT * FROM (
+  SELECT 
+    statvar, 
+    SUBSTR(observation_about, 0, 8) AS observation_about,
+    observation_date, 
+    CONCAT('dcAggregate/', measurement_method) AS measurement_method,
+    population_statvar,
+    SUM(CAST(pop_count AS FLOAT64)) * 100 / SUM(CAST(population AS FLOAT64)) AS percent
```
[the remark this answers is not in the document]
Comment (shvngisingh) [reply scripts/us_cdc/cdc500_state/process.py]: Added

**Comment by pulkeet-wq** [comment]
Add validation_config.json

**Comment by shvngisingh** [comment]
> Add validation_config.json

Added

File: scripts/us_cdc/cdc500_state/validation_config.json [hunk 3]
```diff
@@ -0,0 +1,54 @@
+{
+    "schema_version": "1.0",
+    "rules": [
+        {
+            "rule_id": "check_deleted_records_percent",
+            "description": "Checks that the percentage of deleted points is within the threshold.",
```
Comment (pulkeet-wq) [inline scripts/us_cdc/cdc500_state/validation_config.json]: The above 3 rules are anyways implemented by default. We don't have to add them here again

File: scripts/us_cdc/cdc500_state/README.md [hunk 4]
```diff
@@ -24,7 +24,7 @@ $ python3 process.py
 ### Artifacts
 
 #### Scripts
-[`process.py`](https://github.com/datacommonsorg/data/blob/master//scripts/us_cdc/cdc500_state/process.py)
+[`process.py`](https://github.com/datacommonsorg/data/blob/master/scripts/us_cdc/cdc500_state/process.py)
```
Comment (pulkeet-wq) [inline scripts/us_cdc/cdc500_state/README.md:48]: Please add description of different tables that we are using and the data we are fetching from those tables

File: scripts/us_cdc/cdc500_state/process.py [hunk 5]
```diff
@@ -11,75 +11,133 @@
...
 from absl import logging
 from google.cloud import bigquery
+import pandas as pd
 
+_FLAGS = flags.FLAGS
 _MODULE_DIR = os.path.dirname(os.path.abspath(__file__))
-_OUTPUT_FILE_PATH = os.path.join(_MODULE_DIR + '/CDC500State_Output')
-if not os.path.exists(_OUTPUT_FILE_PATH):
-    os.mkdir(_OUTPUT_FILE_PATH)
+_DEFAULT_OUTPUT_DIR = os.path.join(_MODULE_DIR, 'CDC500State_Output')
```
Comment (pulkeet-wq) [inline scripts/us_cdc/cdc500_state/process.py:24]: what if this path don't exist? How will you handle that scenario?

Is it "OK" to remove the "/" from the name of the folder?

File: scripts/us_cdc/cdc500_state/process.py [hunk 6]
```diff
@@ -11,75 +11,133 @@
...
-    logging.info("Running the query")
-    query_job = client.query(query)
-except Exception as e:
-    logging.fatal(f"Error faced while running the query {e}")
-try:
-    logging.info("Converting to dataframe")
-    results = query_job.to_dataframe()
-except Exception as e:
-    logging.info(f"Error faced while fetching results: {e}")
+def get_query() -> str:
```
Comment (pulkeet-wq) [inline scripts/us_cdc/cdc500_state/process.py]: What is the need of this function? We can directly access the global variable in the script

File: scripts/us_cdc/cdc500_state/process.py [hunk 7]
```diff
@@ -11,75 +11,133 @@
...
+    """Returns the SQL query string for CDC 500 state aggregation."""
+    return QUERY
+
+def run_process(client: bigquery.Client, output_file: str) -> pd.DataFrame:
+    """Executes the BigQuery query and writes the resulting DataFrame to output_file."""
+    logging.info("Running BigQuery aggregation query...")
+    try:
+        query_job = client.query(get_query())
+    except Exception as e:
+        logging.fatal("Failed to submit BigQuery query: %s", e)
```
Comment (pulkeet-wq) [inline scripts/us_cdc/cdc500_state/process.py]: Use logging.error instead of logging.fatal

File: scripts/us_cdc/cdc500_state/process.py [hunk 8]
```diff
@@ -11,75 +11,133 @@
...
+        raise
+
+    logging.info("Fetching query results into dataframe...")
+    try:
+        df = query_job.to_dataframe()
+    except Exception as e:
+        logging.fatal("Failed to fetch query results into dataframe: %s", e)
```
Comment (pulkeet-wq) [inline scripts/us_cdc/cdc500_state/process.py]: Again use logging.error

File: scripts/us_cdc/cdc500_state/process.py [hunk 9]
```diff
@@ -11,75 +11,133 @@
...
+        raise
+
+    output_dir = os.path.dirname(output_file)
+    if output_dir:
+        os.makedirs(output_dir, exist_ok=True)
+    logging.info("Writing %d rows to %s", len(df), output_file)
+    df.to_csv(output_file, index=False)
+    return df
```
Comment (pulkeet-wq) [inline scripts/us_cdc/cdc500_state/process.py]: Why are are returning a df if we are not using it elsewhere. ?

File: scripts/us_cdc/cdc500_state/validation_config.json (hunk 3 above)
Comment (shvngisingh) [reply scripts/us_cdc/cdc500_state/validation_config.json]: Removed the 3 default rules (check_deleted_records_percent, check_missing_refs_count, check_lint_error_count) and retained only the dataset-specific custom validation rules (MAX_VALUE_CHECK, MIN_VALUE_CHECK, and NUM_PLACES_COUNT).

File: scripts/us_cdc/cdc500_state/README.md (hunk 4 above)
Comment (shvngisingh) [reply scripts/us_cdc/cdc500_state/README.md:48]: Added a detailed Data Sources and Processing Methodology section to the README explaining the BigQuery tables (Observation and TimeSeries), the provenances used (CDC500 and CensusACS5YearSurvey), demographic cohort mapping, and the population-weighted state-level aggregation formula.

File: scripts/us_cdc/cdc500_state/process.py (hunk 5 above)
Comment (shvngisingh) [reply scripts/us_cdc/cdc500_state/process.py:24]: 1. Yes, we can remove it. Removing the '/' is standard because os.path.join(_MODULE_DIR, 'CDC500State_Output') is the idiomatic, cross-platform way to construct directory paths (it automatically provides the proper OS separator).
2. For handling non-existent paths, run_process() calls os.makedirs(os.path.dirname(output_file), exist_ok=True) right before saving the CSV, safely creating any missing directories.

File: scripts/us_cdc/cdc500_state/process.py (hunk 6 above)
Comment (shvngisingh) [reply scripts/us_cdc/cdc500_state/process.py]: Agreed, removed the redundant get_query() wrapper function and used the module-level QUERY constant directly in run_process() and unit tests.

File: scripts/us_cdc/cdc500_state/process.py (hunk 7 above)
Comment (shvngisingh) [reply scripts/us_cdc/cdc500_state/process.py]: Updated both error logging statements from logging.fatal to logging.error.

File: scripts/us_cdc/cdc500_state/process.py (hunk 8 above)
Comment (shvngisingh) [reply scripts/us_cdc/cdc500_state/process.py]: removed

File: scripts/us_cdc/cdc500_state/process.py (hunk 9 above)
Comment (shvngisingh) [reply scripts/us_cdc/cdc500_state/process.py]: It was originally returned for testing convenience. Removed return df and updated the return type annotation to None since the function's responsibility is saving the output CSV, and the unit tests verify the generated file.

Comment (pulkeet-wq) [reply scripts/us_cdc/cdc500_state/process.py]: Instead of returning "None"...please return "True" to mark the completion of function successfully

Comment (shvngisingh) [reply scripts/us_cdc/cdc500_state/process.py]: Added

**Review by pulkeet-wq — APPROVED** [review]
